#SCL-8312 fixed check for .flatten emulation

lpiepiora opened the pull request:
This is a PR for https://youtrack.jetbrains.com/issue/SCL-8312, as discussed here: https://youtrack.jetbrains.com/issue/SCL-9562.

Commit:
#SCL-8312 fixed check for .flatten emulation

niktrop commented on src/org/jetbrains/plugins/scala/codeInspection/collections/EmulateFlattenInspection.scala:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  private object identity {
+
```
Method `identity` may also be used with type argument or in the form of postfix expression. You may create an invocation template for `identity` method from `scala.Predef`, like examples in the package object show. It will handle all such cases. It's `unapplySeq` method match qualifier as a first subpattern, you'll just need to ignore it.

lpiepiora replied:
@niktrop Thanks for the tip! I've changed this, and added couple of tests for the new cases, which you pointed out wouldn't work in the original setup. However I have noticed something pretty strange.

Consider this test case:

```scala
  def testSuggests10(): Unit = {
    val selected = s"Set(Set(1), Set(2), Set(3)).${START}flatMap(identity _)$END"
    checkTextHasError(selected, hint, inspectionClass)
    val text = "Set(Set(1), Set(2), Set(3)).flatMap(identity _)"
    val result = "Set(Set(1), Set(2), Set(3)).flatten"
    testFix(text, result, hint)
  }
```

Which is pattern matched by this match (not exactly how I coded, but this is easier to see):

```scala
case seqOfSeqs `.flatMap` (underscore()) =>
```

Where I would expect it to be matched by something like this:

```scala
case seqOfSeqs `.flatMap` (`.identity`(_, underscore())) =>
```

Is my expectation incorrect, or there is some problem with the code. If it's the latter I would like to fix it, but I don't really know where, either PSI is incorrectly generated or the InvocationTemplate is parsing it wrong I guess. I think the PSI is incorrect but I'm not sure.

If I have 

```scala
flatMap(identity(_))
```

The PSI would be like this:

<img width="437" alt="screenshot 2016-04-04 21 22 00" src="https://cloud.githubusercontent.com/assets/105363/14260344/687bb20e-faab-11e5-9861-e06a495c464f.png">

Which is more or less logical especially in regards to `UnderscoreSection` which wraps `_`, and is placed in the argument list.

However when I have code like this:

```scala
flatMap(identity _)
```

The PSI will look like this:

<img width="410" alt="screenshot 2016-04-04 21 25 32" src="https://cloud.githubusercontent.com/assets/105363/14260483/fae6dd8a-faab-11e5-8b19-5167765ba82d.png">

Here the `UnderscoreSection spans from one parenthesis to another, and the identity call is wrapped in the `UnderscoreSection`. Not sure if this is correct, but if so, then this check:

```scala
object underscore {
    def unapply(expr: ScExpression): Boolean = {
      stripped(expr) match {
        case ScParenthesisedExpr(underscore()) => true
        case typed: ScTypedStmt if typed.expr.isInstanceOf[ScUnderscoreSection] => true
        case und: ScUnderscoreSection => true
        case _ => false
      }
    }
  }
```

Seems to be too relaxed in the case `case und: ScUnderscoreSection => true`

Is my thinking correct?

lpiepiora commented:
@niktrop Thanks for the tip! I've changed this, and added couple of tests for the new cases, which you pointed out wouldn't work in the original setup. However I have noticed something pretty strange.

Consider this test case:

``` scala
  def testSuggests10(): Unit = {
    val selected = s"Set(Set(1), Set(2), Set(3)).${START}flatMap(identity _)$END"
    checkTextHasError(selected, hint, inspectionClass)
    val text = "Set(Set(1), Set(2), Set(3)).flatMap(identity _)"
    val result = "Set(Set(1), Set(2), Set(3)).flatten"
    testFix(text, result, hint)
  }
```

Which is pattern matched by this match (not exactly how I coded, but this is easier to see):

``` scala
case seqOfSeqs `.flatMap` (underscore()) =>
```

Where I would expect it to be matched by something like this:

``` scala
case seqOfSeqs `.flatMap` (`.identity`(_, underscore())) =>
```

Is my expectation incorrect, or there is some problem with the code. If it's the latter I would like to fix it, but I don't really know where, either PSI is incorrectly generated or the InvocationTemplate is parsing it wrong I guess. I think the PSI is incorrect but I'm not sure.

If I have 

``` scala
flatMap(identity(_))
```

The PSI would be like this:

<img width="437" alt="screenshot 2016-04-04 21 22 00" src="https://cloud.githubusercontent.com/assets/105363/14260344/687bb20e-faab-11e5-9861-e06a495c464f.png">

Which is more or less logical especially in regards to `UnderscoreSection` which wraps `_`, and is placed in the argument list.

However when I have code like this:

``` scala
flatMap(identity _)
```

The PSI will look like this:

<img width="410" alt="screenshot 2016-04-04 21 25 32" src="https://cloud.githubusercontent.com/assets/105363/14260483/fae6dd8a-faab-11e5-8b19-5167765ba82d.png">

Here the `UnderscoreSection` spans from one parenthesis to another, and the identity call is wrapped in the `UnderscoreSection`. Not sure if this is correct, but if so, then this check:

``` scala
object underscore {
    def unapply(expr: ScExpression): Boolean = {
      stripped(expr) match {
        case ScParenthesisedExpr(underscore()) => true
        case typed: ScTypedStmt if typed.expr.isInstanceOf[ScUnderscoreSection] => true
        case und: ScUnderscoreSection => true
        case _ => false
      }
    }
  }
```

Seems to be too relaxed in the case `case und: ScUnderscoreSection => true`

Is my thinking correct?

niktrop commented:
Good catch! It's too relaxed indeed. 

`ScUnderscoreSection` is generated by our parser in two cases. First is placeholder for a parameter in the short form of function expression (like in `seq.map(_ + 1)`) and second is a method value. Probably it would be better to separate them to different classes. But now you can distinguish them with `bindingExpr` method. It returns reference to the method in the case of method value.

You can check extractor for a function expression `contains` in `FilterOtherContainsInspection.scala`, it does similar thing.

niktrop commented on src/org/jetbrains/plugins/scala/codeInspection/collections/EmulateFlattenInspection.scala:
```diff
@@ -0,0 +1,39 @@
+    }
+
+  private object identityOperation {
+    def unapply(expr: ScExpression): Boolean = stripped(expr) match {
```
I would use syntax for infix patterns here: `_'.identity' (underscore())` , but it's up to you.

lpiepiora replied:
Good point, I will change it, because it actually reads more natural I think, since the first param is really a qualifier.

Commit:
#SCL-8312 make underscore check more strict

lpiepiora commented:
I've changed the code as you've suggested to check for `bindingExpr`. 

I think I understand now the intention. It wasn't clear to me (maybe still is a bit unclear), why `ScUnderscoreSection` once wraps the method value, and once it just wraps underscore. However I guess, that when it wraps the method value, then the `bindingExpr` is set to that method. Given that assumption I have changed the check in `underscore()`, so it does not pattern match against the example from my previous comment, by checking if `bindingExpr` is empty.

niktrop commented:
Thanks for contributing! I merged your changes manually.
